accepted/2021/system-drawing-win-only/system-drawing-win-only.md

```diff
@@ -0,0 +1,183 @@
+libraries, which makes shipping even more challenging.
+
+We've also hit some memory leaks or double free'd memory that weren't fixed in
+`libgdiplus` and that we had to work around because they were potential security
+issues. This makes it harder as we are basically on our “own” for those security
+issues. `libgdiplus` is a library we don't own, but that we must spend time
+investigating issues and proposing fixes.
+
+We have noticed that `System.Drawing.Common` is used on cross-platform mostly for
+image manipulation like QR code generators and text rendering. We haven't
+noticed heavy graphics usage as our cross-platform graphics support is very
+incomplete. For example, right now some components crash on macOS because the
+native APIs we use have been deprecated and are no longer included with recent
+macOS releases. We have never gotten a report about that crash; we discovered it
+ourselves while making `System.Drawing.Common` trim compatible.
+
+The usages we see of `System.Drawing.Common` in non-Windows environments are
+typically well supported with SkiaSharp and ImageSharp.
+
+Thus, we believe it is not worth investing in making `System.Drawing.Common` work
+on non-Windows platforms and instead propose we only keep evolving it in the
+context of Windows Forms and GDI+.
+
+## Requirements
+
+### Goals
+
+* Consumers of `System.Drawing.Common` will be informed that starting with .NET
+  6, it's only supported on Windows, via the platform compatibility analyzer
+* Cross-platform consumers have the ability to turn on a compatibility switch to
+  continue using `System.Drawing.Common` via `libgdiplus`
+
+### Non-Goals
+
+* Closing the gap of `System.Drawing.Common` for non-Windows platforms or
+  bundling `libgdiplus` as part of the .NET 6 installer
+
+## Stakeholders and Reviewers
+
+* Libraries team
+* ML.NET
+* Xamarin team
+
+## Design
+
+In the past we decided to include the following disclaimer in our docs:
+
+>   The `System.Drawing.Common` NuGet package works on Windows, Linux, and macOS.
+>   However, there are some platform differences. On Linux and macOS, the GDI+
+>   functionality is implemented by the `libgdiplus` library. This library is not
+>   installed by default in most Linux distributions and doesn't support all the
+>   functionality of GDI+ on Windows on macOS. There are also platforms where
+>   `libgdiplus` is not available at all. To use types from the
+>   `System.Drawing.Common` package on Linux and macOS, you must install
+>   `libgdiplus` separately. For more information, see Install .NET on Linux or
+>   Install .NET on macOS.
+>
+>   If you can't use System.Drawing with your application, recommended
+>   alternatives include ImageSharp, SkiaSharp, and Windows Imaging Components.
+
+However, this disclaimer doesn't feel clear enough to steer people towards using
+other graphics libraries.
+
+The proposal is to do the following for .NET 6:
```

Comment:
What is going to be the plan for .NET 7+? Are we going to keep the libgdiplus in increasingly more broken state forever?

Can we be more aggressive in .NET 6 to be done with this? Make it throw unconditionally, and tell people who need the Unix support to use the older version of the package.

Reply:
> Can we be more aggressive in .NET 6 to be done with this? Make it throw unconditionally

I don't think we should do that until we at least migrate our own things. ML.NET suddenly throwing an exception when you upgrade to .NET 6 doesn't sound like a great experience to me.

Reply:
I support this, the library hasn't evolved for Unix in any release honestly, so using a previous version of the package for Unix support makes sense and there aren't any features to be lost.

Reply:
Sounds good.

@safern do you want to update the doc? You can just push to this branch.

Reply:
Yup.

Reply:
Do we really have to put a poison pill into the package? Instead can't we just say it targets `windows`, and users will get warnings from the analyzer automatically when they try using it cross-plat. If they don't care, they can suppress the warnings. And we can make the statement that we don't support it on non-Windows, your mileage may vary. If it continues to work for you, great. But if you run into issues, we won't be fixing any bugs or new features going forward.

Basically, it is deprecated. Use at your own risk.

Reply:
While I understand the concern, how would that be different from just recommending to pin the package version to a specific version? The fact that you communicate that we recommend a package that is potentially out of support for both Windows and non-Windows?

Reply:
> your mileage may vary. If it continues to work for you, great. But if you run into issues, we won't be fixing any bugs or new features going forward

Feels to me we need to be more prescriptive. We led customers to use this library on Unix. They had every reason to believe it would be supported on the same lifecycle, but it will not be, potentially opening stability or security concerns. In my mind we should say essentially

* when you target your app to .NET 6 or later, any use of System.Drawing on Unix platforms will fail.
* to keep existing apps working while adopting .NET 6, you can pin to the previous version of System.Drawing as a temporary measure <instructions>. you should pin only for Unix deployed applications, so that Windows deployed applications recieve the latest updates and bug fixes contained in the latest System.Drawing.
* however long term, we strongly encourage migrating such code to an alternative library on Unix. Alternatives include SkiaSharp and ImageSharp. These work well on both Windows and Unix.
* as a reminder, usage of System.Drawing on Windows remains fully supported as it has always been.
?

Reply:
We can afford to be more aggressive in our deprecation than BinaryFormatter, which became engrained throughout our platform and ecosystem over 20 years.

Reply:
At least users get bug fixes/improvements that are common across platforms. For example, nullable annotations and ILLink annotations.

A concrete advantage: using `System.Drawing.Common` v5 in a trimmed app produces ILLink warnings from inside System.Drawing.Common that developers can't fix. Using v6 does not.

Reply:
After giving it a second thought. Given that 5.0.0 package will be out of support 3 months after the 6.0 release, I think it makes sense to do the original proposal which is:

1. Mark the assembly as windows only so that the platform analyzer warns
2. Throw PNSE by default when loading libgdiplus with a link to the aka.ms speaking about unix not being officially supported by us.
3. Providing a compat switch to opt-in to the unix implementation and don't throw when that is enabled.
4. In 7.0 nuke the unix implementation entirely.

The reason I thin we should follow that plan is because there is no way to cross target by RID so you can not say on a project, use this package version for windows but this package version for unix, so all these libraries that depend on our package will be out of support on windows in 3 months after the release and I believe we should give them more time to adjust, with the constraint that _Unix is no longer supported by us and we will not fix any unix bugs in our product nor libgdiplus_.

Thoughs, @eerhardt @danmoseley @terrajobst @jkotas

Reply:
Sounds good to me.

Reply:
That sounds good to me given 5.0 will be going out of support next year. Opt-in to Unix.

Reply:
I like the plan.

Reply:
+1